File: app/internal/utils/tagcolors.go
package utils

// TagColorOption is a selectable tag color from the palette.
type TagColorOption struct {
	Key  string
	Hex  string
	Name string
}

var tagPalette = []TagColorOption{
	{Key: "blue", Hex: "#3b82f6", Name: "Blue"},
	{Key: "indigo", Hex: "#6366f1", Name: "Indigo"},
	{Key: "violet", Hex: "#8b5cf6", Name: "Violet"},
	{Key: "emerald", Hex: "#10b981", Name: "Emerald"},
	{Key: "teal", Hex: "#14b8a6", Name: "Teal"},
	{Key: "cyan", Hex: "#06b6d4", Name: "Cyan"},
	{Key: "amber", Hex: "#f59e0b", Name: "Amber"},
	{Key: "orange", Hex: "#f97316", Name: "Orange"},
	{Key: "rose", Hex: "#f43f5e", Name: "Rose"},
	{Key: "red", Hex: "#ef4444", Name: "Red"},
	{Key: "slate", Hex: "#64748b", Name: "Slate"},
}

var tagColorHex = map[string]string{}

func init() {
	for _, c := range tagPalette {
		tagColorHex[c.Key] = c.Hex
	}
}

// DefaultTagColor returns the default palette key for new tags.
func DefaultTagColor() string {
	return "blue"
}

// IsValidTagColor reports whether key is in the palette.
func IsValidTagColor(key string) bool {
	if key == "" {
		return false
	}
	_, ok := tagColorHex[key]
	return ok
}

// NormalizeTagColor returns a valid palette key, defaulting invalid values to blue.
func NormalizeTagColor(key string) string {
	if IsValidTagColor(key) {
		return key
	}
	return DefaultTagColor()
}

File: frontend/src/lib/tag-colors.ts
export type TagColorOption = {
  key: string;
  hex: string;
  name: string;
};

export const TAG_COLORS: TagColorOption[] = [
  { key: "blue", hex: "#3b82f6", name: "Blue" },
  { key: "indigo", hex: "#6366f1", name: "Indigo" },
  { key: "violet", hex: "#8b5cf6", name: "Violet" },
  { key: "emerald", hex: "#10b981", name: "Emerald" },
  { key: "teal", hex: "#14b8a6", name: "Teal" },
  { key: "cyan", hex: "#06b6d4", name: "Cyan" },
  { key: "amber", hex: "#f59e0b", name: "Amber" },
  { key: "orange", hex: "#f97316", name: "Orange" },
  { key: "rose", hex: "#f43f5e", name: "Rose" },
  { key: "red", hex: "#ef4444", name: "Red" },
  { key: "slate", hex: "#64748b", name: "Slate" },
];

const colorHexMap = Object.fromEntries(TAG_COLORS.map((c) => [c.key, c.hex]));

// tagColorHex resolves a palette key to its hex value, defaulting to slate.
export function tagColorHex(key: string): string {
  return colorHexMap[key] ?? colorHexMap.slate;
}

// defaultTagColor returns the default palette key for new tags.
export function defaultTagColor(): string {
  return "blue";
}
